src/interface/variable_pack.hpp

```diff
@@ -37,34 +43,98 @@ class ParticleVariable;
 // some convenience aliases
 namespace vpack_types {
 template <typename T>
-using VarList = std::forward_list<std::shared_ptr<CellVariable<T>>>;
-template <typename T>
 using SwarmVarList = std::forward_list<std::shared_ptr<ParticleVariable<T>>>;
+
 // Sparse and/or scalar variables are multiple indices in the outer view of a pack
 // the pairs represent interval (inclusive) of those indices
 using IndexPair = std::pair<int, int>;
+
 // Flux packs require a set of names for the variables and a set of names for the strings
 // and order matters. So StringPair forms the keys for the FluxPack cache.
 using StringPair = std::pair<std::vector<std::string>, std::vector<std::string>>;
+
 } // namespace vpack_types
 
-// using PackIndexMap = std::map<std::string, vpack_types::IndexPair>;
+// helper class to make lists of variables with labels
+template <typename T>
+class VarListWithLabels {
+ public:
+  VarListWithLabels<T>() = default;
+
+  // Adds a variable to the list if one of the following is true:
+  // a) The variable is not sparse
+  // b) The set of sparse_ids is empty
+  // c) The sparse id of the variable is contained in the set of sparse_ids
+  void Add(const std::shared_ptr<CellVariable<T>> &var,
+           const std::unordered_set<int> &sparse_ids = {}) {
+    if (!var->IsSparse() || sparse_ids.empty() ||
+        (sparse_ids.count(var->GetSparseID()) > 0)) {
+      vars_.push_back(var);
+      // The label here is used in the key for variable pack caching, so it has to include
+      // the allocation status of the variable, since we cannot reuse a variable pack if
+      // one of its variables has a changed allocation status
+      labels_.push_back(var->label() + (var->IsAllocated() ? "_A" : ""));
```

Comment:
By adding the allocation status to the label used to construct the key, we won't have to invalidate all caches whenever an allocation status changes. We might have a number of packs that don't contain sparse variables and so those packs don't need to be thrown out when the allocation status of a sparse variable changes. It is true, though, that every time an allocation status changes, we could just go through the cached packs and throw out the ones involving the field(s) whose allocation status has changed. But since that happens at the `CellVariable` level, we'd have to make the `MeshBlockData` explicitly aware of that, which might reduce encapsulation and increase code complexity. On the other hand, including the allocation status in the key means we have to periodically clear the cache of invalid packs. We could do this by assigning a lifetime to them that resets every time it's accessed and then clear the packs that haven't been accessed in a while.

Yes, the key won't be the same as the vector of strings of requested names. But even without this, they're not guaranteed to be the same anyway. If I request a pack with `{"varA", "varB", "varC"}`, but `varB` doesn't exist, the key will be `{"varA", "varC"}` (assuming they're all dense). If `varC` was sparse and I don't specify any indices, they key would be something like `{"varA", "varC_0", "varC_3", "varC_42"}`, assuming `varC` has sparse IDs 0, 3, and 42. So the user can never assume that the string of variable names they pass to `Pack*` will be the same as the key of the pack. We also discussed sorting the variables in the pack, but I don't think we're doing this yet. However, the key that is returned through the optional output arguments of the `Pack*` functions will always be exactly the internal key.

So at the end of the day, I think I have a preference for keeping the allocation status in the key. I don't see any downsides and I think it would make managing the cache easier (but that remains to be seen). In any case, I think it might help with debugging if the pack keys include the allocation status.

Reply:
Giving packs a lifetime seems kind of gross. Is that definitely necessary?

Reply:
I think we do need a mechanism to delete packs that are no longer valid because they contain a sparse field whose allocation status has changed. My plan was to have the allocation status be part of the cache key, which would ensure that we only use a cached pack if it was made with the exact same allocation statuses as we have now. However, in that case we'll need a mechanism to remove old packs (otherwise the `CellVariable` data that is now no longer needed won't be freed, since there is still a copy of the view floating around). I thought we could introduce an expiration for packs, so if a pack hasn't been accessed in, say, 10 time steps, it gets removed from the cache. I do agree, however, that it adds some complexity.

The other way I was thinking of doing it so to basically check the cached packs whenever an allocation status changes and throw out the ones that involve the fields whose allocation status changed. However, I think this would be more complex than the previous idea and so I'm not very happy with this approach.

However, after thinking about this some more, I think I have a better solution. I propose we remove the allocation status from the cache key, but store it separately as a `std::vector<bool>` inside the pack. When a pack is requested, we have to generate the key anyway and we'll just generate the `std::vector<bool>` with allocation statuses at the same time. If a pack with the given key already exists in the cache, we check that the allocation statuses match. If yes, we return the cached pack, otherwise we overwrite the cache entry with a new pack. This should be pretty straight forward to implement and I'll add it to this PR unless there are any objections.

Reply:
I like your new solution @jlippuner ! Only one thing to be careful about is that this works correctly for both `VariablePack` and `MeshBlockPack`.